Service's sendable messages added to Client-Registry

trushton opened the pull request:
The `SERVICE_MESSAGES` environment variable from `exo-run` will now contain the `sends` field of each service's `service.yml`. This update for ExoCom will store that information in the `client-registry`. When a message is to be sent, verification will be done to make sure that the service had that message in its `service.yml` on start-up by reading from the `sends` field of `client-registry`. 
@kevgo @hugobho @martinjaime 
Addresses [/Originate/exosphere-sdk#171](https://github.com/Originate/exosphere-sdk/issues/171)

hugobho commented on features/register-service.feature:
```diff
@@ -10,7 +10,14 @@ Feature: Manage new instances of services
+        "name": "users-service"
+      }
+    ]
+    """
```
I don't think we need routing information at all for this feature test. It's just testing that ExoCom can register services, it doesn't send any messages (which is the only time we need to look up service messages).

hugobho replied:
We should just be able to write `Given an ExoCom instance` and pass it an empty `service-messages` object.

kevgo commented on features/broadcasting-messages.feature:
```diff
@@ -25,3 +39,9 @@ Feature: Broadcasting messages
+  # ERROR HANDLING
+  Scenario: broadcasting an invalid message
+    When the "web" service sends "users.get-SSN"
+    Then ExoCom signals the error "invalid message"
```
You should also say here that ExoCom does not broadcast the message.

kevgo commented on features/translating-messages.feature:
```diff
@@ -11,7 +11,21 @@ Feature: Translating messages
+      },
+      {
+        "name": "tweets",
+        "receives": ["text-snippets.create"],
```
I think this route must also contain the `namespace` of the service, so that ExoCom knows to translate `tweets.create` to `text-snippets.create`.

trushton replied:
Should this be included in this PR, or separated out into its own?

kevgo replied:
Up to you, as long as you do it  :)

kevgo commented on src/websocket-subsystem/websocket-subsystem.ls:
```diff
@@ -58,8 +58,10 @@ class WebSocketSubsystem extends EventEmitter
     if request-data.name is "exocom.register-service"
       @on-registration-receive request-data.payload, websocket
-    else
+    else if @valid-sender request-data.sender, request-data.name
```
I think a more intuitive pattern is to check for exceptions first (like if this is not a valid sender), and then at the end do the normal case, i.e. call `on-message-receive`. This way we can add more edge cases in the future. Also, you might want to use pattern matching here.

```ls
switch
  | request-data.name is "exocom.register-service"          =>  @on-registration-receive request-data.payload, websocket
  | @invalid-sender request-data.sender, request-data.name  =>  @emit 'error', "Service '#{request-data.sender}' attempted to send invalid message '#{request-data.name}'"
  | otherwise                                               =>  @on-message-receive request-data

kevgo commented on src/websocket-subsystem/websocket-subsystem.ls:
```diff
@@ -131,4 +133,9 @@ class WebSocketSubsystem extends EventEmitter
 
 
+  valid-sender: (sender, message) ->
+    @exocom.client-registry.sends[sender] |> (.includes message)
```
Obey the law of Demeter. Deciding whether a sender is allowed to send a message is a responsibility of the client registry, not the websocket subsystem. The latter should not have any knowledge how the former is implemented internally.

```ls
valid-sender: (sender, message) ->
  @exocom.client-registry.can-send sender, message

trushton commented:
@kevgo Addressed comments, merged routing info from the `SERVICE_MESSAGES` environment variable into one object in `client-registry`. Does this affect the current `@clients` and `@routes` objects that were already there?
